File: java/src/test/java/io/cucumber/query/QueryTest.java
package io.cucumber.query;

import io.cucumber.messages.types.Envelope;
import io.cucumber.messages.types.TestCaseFinished;
import io.cucumber.messages.types.TestCaseStarted;
import io.cucumber.messages.types.Timestamp;
import org.junit.jupiter.api.Test;

import java.util.UUID;
import java.util.stream.Stream;

import static org.assertj.core.api.Assertions.assertThat;

class QueryTest {

    final Repository repository = Repository.builder().build();
    final Query query = new Query(repository);

    @Test
    void retainsInsertionOrderForTestCaseStarted() {
        TestCaseStarted a = new TestCaseStarted(0L, randomId(), randomId(), "main", new Timestamp(1L, 0));
        TestCaseStarted b = new TestCaseStarted(0L, randomId(), randomId(), "main", new Timestamp(1L, 0));
        TestCaseStarted c = new TestCaseStarted(0L, randomId(), randomId(), "main", new Timestamp(1L, 0));

        Stream.of(a, b, c)
                .map(Envelope::of)
                .forEach(repository::update);

        assertThat(query.findAllTestCaseStarted()).containsExactly(a, b, c);
    }

    @Test
    void omitsTestCaseStartedIfFinishedAndWillBeRetried() {
        TestCaseStarted a = new TestCaseStarted(0L, randomId(), randomId(), "main", new Timestamp(0L, 0));
        TestCaseFinished b = new TestCaseFinished(a.getId(), new Timestamp(0L, 0), true);
        TestCaseStarted c = new TestCaseStarted(0L, randomId(), randomId(), "main", new Timestamp(0L, 0));
        TestCaseFinished d = new TestCaseFinished(c.getId(), new Timestamp(0L, 0), false);

        Stream.of(a, c)
                .map(Envelope::of)
                .forEach(repository::update);
        Stream.of(b, d)
                .map(Envelope::of)
                .forEach(repository::update);

        assertThat(query.findAllTestCaseStarted()).containsExactly(c);
        assertThat(query.countTestCasesStarted()).isEqualTo(1);
    }

    private static String randomId() {
        return UUID.randomUUID().toString();
    }

}

File: dotnet/QueryTest/QueryTest.cs
using Io.Cucumber.Messages.Types;
using Io.Cucumber.Query;

namespace QueryTest
{
    [TestClass]
    public class QueryTest
    {
        private readonly Repository _repository;
        private readonly Query _query;

        public QueryTest()
        {
            _repository = new Repository();
            _query = new Query(_repository);
        }

        [TestMethod]
        public void RetainsInsertionOrderForTestCaseStarted()
        {
            var a = new TestCaseStarted(0L, RandomId(), RandomId(), "main", new Timestamp(1L, 0L));
            var b = new TestCaseStarted(0L, RandomId(), RandomId(), "main", new Timestamp(1L, 0L));
            var c = new TestCaseStarted(0L, RandomId(), RandomId(), "main", new Timestamp(1L, 0L));

            foreach (var tcs in new[] { a, b, c })
                _repository.UpdateTestCaseStarted(tcs);

            var result = _query.FindAllTestCaseStarted();
            CollectionAssert.AreEqual(new[] { a, b, c }, result.ToArray());
        }

        [TestMethod]
        public void OmitsTestCaseStartedIfFinishedAndWillBeRetried()
        {
            var a = new TestCaseStarted(0L, RandomId(), RandomId(), "main", new Timestamp(0L, 0L));
            var b = new TestCaseFinished(a.Id, new Timestamp(0L, 0L), true);
            var c = new TestCaseStarted(0L, RandomId(), RandomId(), "main", new Timestamp(0L, 0L));
            var d = new TestCaseFinished(c.Id, new Timestamp(0L, 0L), false);

            _repository.UpdateTestCaseStarted(a);
            _repository.UpdateTestCaseStarted(c);
            _repository.UpdateTestCaseFinished(b);
            _repository.UpdateTestCaseFinished(d);

            var result = _query.FindAllTestCaseStarted();
            Assert.AreEqual(1, result.Count());
            Assert.AreEqual(c, result.ToArray()[0]);
            Assert.AreEqual(1, _query.TestCasesStartedCount);
        }

        private static string RandomId() => Guid.NewGuid().ToString();
    }
}
